File: dev_core/src/utils/AppError.js
const ErrorCodes = require('../constants/errorCodes')

/**
 * 应用错误类
 * 用于统一错误处理
 */
class AppError extends Error {
  constructor(errorCode, statusCode = 400, options = {}) {
    const normalizedOptions = options && typeof options === 'object' ? options : {}
    const message = typeof normalizedOptions.message === 'string' ? normalizedOptions.message : ''

    // Error.message 仍保留，以便日志与调试栈可读。
    super(message || '应用错误')

    // 兼容历史字段（errorCode），并补充新契约字段（code/msg）。
    this.errorCode = errorCode
    this.code = ErrorCodes.toPublicCode(errorCode)
    this.msg = message
    this.statusCode = statusCode
    this.options = normalizedOptions
    this.name = 'AppError'

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AppError)
    }
  }
}

module.exports = AppError

File: data_service/internal/errors/app_error.go
package errors

import "fmt"

// AppError 表示 data_service 的业务错误。
type AppError struct {
	Code       ErrorCode
	Message    string
	StatusCode int
	Err        error
}

// NewAppError 创建一个业务错误。
func NewAppError(code ErrorCode, statusCode int, message string) *AppError {
	return &AppError{
		Code:       code,
		Message:    message,
		StatusCode: statusCode,
	}
}

// WrapAppError 在保留原始错误的同时创建业务错误。
func WrapAppError(code ErrorCode, statusCode int, message string, err error) *AppError {
	return &AppError{
		Code:       code,
		Message:    message,
		StatusCode: statusCode,
		Err:        err,
	}
}

// Error 返回错误的文本表示。
func (e *AppError) Error() string {
	if e == nil {
		return ""
	}
	if e.Err != nil {
		return fmt.Sprintf("%s: %v", e.Message, e.Err)
	}
	return e.Message
}

// Unwrap 返回被包装的原始错误。
func (e *AppError) Unwrap() error {
	if e == nil {
		return nil
	}
	return e.Err
}
